### PR TITLE
ar71xx: add support for Fritz!Box 4020

### PR DESCRIPTION
**Pull request by blocktrron**
This adds support for the AVM Fritz!Box 4020 WiFi-router.

Tested and working:
 - Ethernet (LAN + WAN)
 - WiFi (correct MAC)
 - Installation via EVA bootloader
 - OpenWRT sysupgrade
 - Buttons
 - LEDs

Not working:
 - USB port

Installation via EVA:
In the first seconds after Power is connected, the bootloader will listen for FTP connections on 169.254.157.1 (Might also be 192.168.178.1). 
Firmware can be uploadedlike following:
```
  ftp> quote USER adam2
  ftp> quote PASS adam2
  ftp> binary
  ftp> debug
  ftp> passive
  ftp> quote MEDIA FLSH
  ftp> put openwrt-sysupgrade.bin mtd1
```

Note that this procedure might take up to two minutes. After transfer is complete you need to powercycle the device to boot OpenWRT.

Regarding the USB Port: Both Root Hubs are detected as having 0 Ports:
```
[    3.670807] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    3.723267] usbcore: registered new interface driver usbfs
[    3.729058] usbcore: registered new interface driver hub
[    3.734616] usbcore: registered new device driver usb
[    3.744181] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    3.758357] SCSI subsystem initialized
[    3.766026] ehci-platform: EHCI generic platform driver
[    3.771548] ehci-platform ehci-platform.0: EHCI Host Controller
[    3.777708] ehci-platform ehci-platform.0: new USB bus registered, assigned bus number 1
[    3.788169] ehci-platform ehci-platform.0: irq 48, io mem 0x1b000000
[    3.816647] ehci-platform ehci-platform.0: USB 2.0 started, EHCI 0.00
[    3.824001] hub 1-0:1.0: USB hub found
[    3.828219] hub 1-0:1.0: config failed, hub doesn't have any ports! (err -19)
[    3.835825] ehci-platform ehci-platform.1: EHCI Host Controller
[    3.842009] ehci-platform ehci-platform.1: new USB bus registered, assigned bus number 2
[    3.852481] ehci-platform ehci-platform.1: irq 49, io mem 0x1b400000
[    3.886631] ehci-platform ehci-platform.1: USB 2.0 started, EHCI 0.00
[    3.894011] hub 2-0:1.0: USB hub found
[    3.898190] hub 2-0:1.0: config failed, hub doesn't have any ports! (err -19)
[    3.908928] usbcore: registered new interface driver usb-storage
[    3.915634] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
```

### CODE REVIEW

File: target/linux/ar71xx/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -299,6 +299,12 @@ fritz300e)
 	ucidef_set_led_rssi "rssimediumhigh" "RSSIMEDIUMHIGH" "$board:green:rssi3" "wlan0" "60" "100"
 	ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "$board:green:rssi4" "wlan0" "80" "100"
 	;;
+fritz4020)
+	ucidef_set_led_default "power" "Power" "$board:green:power" "1"
+	ucidef_set_led_netdev "lan" "LAN" "$board:green:lan" "eth1"
+	ucidef_set_led_netdev "wan" "WAN" "$board:green:wan" "eth0"
+	ucidef_set_led_wlan "wlan" "WLAN" "$board:green:wlan" "phy0tpt"
+	;;
 dap-1330-a1)
```
Comment (blogic) [inline target/linux/ar71xx/base-files/etc/board.d/01_leds]: alphabetical ordering please

File: target/linux/ar71xx/base-files/lib/ar71xx.sh [hunk 2]
```diff
@@ -712,6 +712,9 @@ ar71xx_board_detect() {
 	*"FRITZ!WLAN Repeater 300E")
 		name="fritz300e"
 		;;
+	*"FRITZ!Box 4020")
```
Comment (blogic) [inline target/linux/ar71xx/base-files/lib/ar71xx.sh]: alphabetical ordering please

File: target/linux/ar71xx/files/arch/mips/ath79/dev-eth.c [hunk 3]
```diff
@@ -1175,6 +1175,14 @@ void __init ath79_parse_ascii_mac(char *mac_str, u8 *mac)
 	}
 }
 
+void __init ath79_extract_mac_reverse(u8 *ptr, u8 *out)
+{
+	int i;
+	for(i = 0; i < ETH_ALEN; i++) {
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/dev-eth.c]: blank like before between variable declaration and code block

Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/dev-eth.c]: for <space> (....

File: target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+#define FRITZ4020_74HC_GPIO_BASE		32
+#define FRITZ4020_74HC_GPIO_LED_LAN	(FRITZ4020_74HC_GPIO_BASE + 0)
+#define FRITZ4020_74HC_GPIO_LED_INFO_RED	(FRITZ4020_74HC_GPIO_BASE + 1)
+#define FRITZ4020_74HC_GPIO_LED_POWER		(FRITZ4020_74HC_GPIO_BASE + 2)
+#define FRITZ4020_74HC_GPIO_LED_WLAN		(FRITZ4020_74HC_GPIO_BASE + 3)
+#define FRITZ4020_74HC_GPIO_LED_WAN		(FRITZ4020_74HC_GPIO_BASE + 4)
+#define FRITZ4020_74HC_GPIO_USB_RST		  (FRITZ4020_74HC_GPIO_BASE + 5)
+#define FRITZ4020_74HC_GPIO_LED_INFO		  (FRITZ4020_74HC_GPIO_BASE + 6)
+
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c]: lots of space vs tab errors here

File: target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+#define FRITZ4020_GPIO_BTN_WPS	2
+#define FRITZ4020_GPIO_BTN_WLAN	21
+#define FRITZ4020_KEYS_POLL_INTERVAL	 	20 /* msecs */
+#define FRITZ4020_KEYS_DEBOUNCE_INTERVAL	(3 * FRITZ4020_KEYS_POLL_INTERVAL)
+
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c]: and here

File: target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+#define FRITZ4020_PHY_MASK 0
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c]: should this not be something like 1 or 4 ?

File: target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+		.size		        = 0x0080000,
+		.mask_flags	    = MTD_WRITEABLE,
+	}, {
+		.name           = "tffs (2)",
+		.offset         = 0x0f80000,
+		.size           = 0x0080000,
+		.mask_flags     = MTD_WRITEABLE,
+	}
+};
+
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c]: tabs vs spaces here aswell

File: target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+	/* Register LED shift-register */
+	spi_register_board_info(fritz4020_spi_info,
+				ARRAY_SIZE(fritz4020_spi_info));
+	platform_device_register(&fritz4020_spi_device);
+
+	/* Register GPIO buttons */
+	ath79_register_gpio_keys_polled(-1, FRITZ4020_KEYS_POLL_INTERVAL,
+										ARRAY_SIZE(fritz4020_gpio_keys),
+										fritz4020_gpio_keys);
```
Comment (blogic) [inline target/linux/ar71xx/files/arch/mips/ath79/mach-fritz4020.c]: few tabs too many here

**Comment by blogic** [comment]
looking good ! just a few minor nitpicks ...

**Comment by blocktrron** [comment]
@blogic Thanks for your feedback! I've corrected the mistakes you pointed out, anything else?

**Comment by blogic** [comment]
@blocktrron 
Please use git commit --amend to add changes to your commit. Use git push --force to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

**Comment by blogic** [comment]
merged, Thanks ! if you have time, consider moving this board over to the new ath79 target

**Pull request closed without merging**